### DISCUSSION TITLE
HTML file instead of html in js

### DISCUSSION THREAD

**Discussion by thegamerx1** [vikejs/vike#1170] [Discussion] [answered]
Is it possible to use an HTML file instead of raw HTML in JS?
I can't get escapeInject to work with a non template string

```ts
let html = readFileSync("renderer/index.html").toString();
...
return escapeInject(html.replace("%HTML%", viewHtml._escaped));
```

**Comment by lourot** [comment]
This nasty workaround works (tested locally):

```js
const html = `<!DOCTYPE html>
<html>
  <body>
    <div id="page-view">%HTML%</div>
  </body>
</html>`; // or readFileSync("renderer/index.html").toString();

const templateStrings = html.split('%HTML%');
return escapeInject(templateStrings, viewHtml);
```

@brillout I would be up for providing a function version of `escapeInject()`. This looks like a valid use case to me. What do you think?

**Reply by brillout** [reply]
Alternatively we can expose `escapeHtml()`. The neat thing about wrapping the whole HTML string `dangerouslySkipEscape()` is that the user has then full control over the HTML generation and doesn't need to fiddle around with the whole `escapeInject` thing. The purpose of the template string tag `escapeInject` is to have a declarative way to define HTML in JS. For programmatic control I think I prefer exposing `escapeHtml` as it gives gives the user full control/responsibility.

```js
return dangerouslySkipEscape(`This can be any string the user wants but the user has then has
to be carefull to escape injections himself`);
```

**Comment by brillout** [comment, marked as answer]
You should do this instead:

```diff
- return escapeInject(html.replace("%HTML%", viewHtml._escaped));
+ return dangerouslySkipEscape(html.replace("%HTML%", viewHtml._escaped));
```

@AurelienLourot I believe it's more of an issue with the docs. It's fairly common that users don't come up with the idea of wrapping the entire HTML string with `dangerouslySkipEscape()`.

**Reply by brillout** [reply]
And if `viewHtml` is the HTML of `Page` then do this:

```js
return dangerouslySkipEscape(html.replace("%HTML%", viewHtml));
```

**Reply by lourot** [reply]
Yes in the [documentation](https://vike.dev/escapeInject) it feels like `escapeInject` != `dangerouslySkipEscape`:

> The `escapeInject` string template tag sanitizes HTML
> [...]
> The `dangerouslySkipEscape(str)` function injects the string `str` as-is *without* sanitizing.

so I wouldn't come up with the idea of using them interchangeably

**Reply by lourot** [reply]
also it feels like what we want to do here has nothing to do with escaping, because all we want is wrapping the (trusted?) output of react/vue/solid inside a trusted HTML document we have written ourselves, so I can't really wrap my head around the escaping considerations in this context and why I would have to use any of these functions

**Reply by brillout** [reply]
The docs are actually accurate here, that's why I'm thinking the docs aren't clear enough. Maybe the docs should provide some minimal code snippet examples showcasing what `escapeInject` and `dangerouslySkipEscape()` do and return.

>  I can't really wrap my head around the escaping considerations in this context and why I would have to use any of these functions

Yes for OP's example there is no need to escape anything, thus the solution is to wrap everything with `dangerouslySkipEscape()`. Vike doesn't know that it's safe: that's why the function is called `dangerouslySkipEscape()`.
